c++/src/kj/common.h

```diff
@@ -1292,11 +1291,68 @@ inline T* readMaybe(const Maybe<T&>& maybe) { return maybe.ptr; }
 
 template <typename T>
 inline T* readMaybe(T* ptr) { return ptr; }
-// Allow KJ_IF_MAYBE to work on regular pointers.
+// Allow KJ_IF_SOME to work on regular pointers.
+
+#if defined(KJ_DEPRECATE_KJ_IF_MAYBE) && KJ_DEPRECATE_KJ_IF_MAYBE
+[[deprecated("KJ_IF_MAYBE is deprecated and will be removed. Please use KJ_IF_SOME instead.")]]
+constexpr int KJ_IF_MAYBE_IS_DEPRECATED = 0;
+#define KJ_DEPRECATE_KJ_IF_MAYBE_STMT \
+    [[maybe_unused]] int KJ_UNIQUE_NAME(deprecation) = ::kj::_::KJ_IF_MAYBE_IS_DEPRECATED
+#else
+#define KJ_DEPRECATE_KJ_IF_MAYBE_STMT
+#endif
+// Before KJ_IF_SOME, we used KJ_IF_MAYBE, which does exactly the same thing as KJ_IF_SOME, but
+// provides a guaranteed-non-null pointer to the wrapped object instead of a reference. This has a
+// tendency to allow authors to inadvertently use pointers where they mean to use references. For
+// example, in `KJ_IF_MAYBE(obj, maybe) { KJ_LOG(INFO, obj); }` would stringify and log the address
+// of the object living in `maybe`, rather than the object itself as perhaps intended.
+//
+// Due to this footgun, we've deprecated KJ_IF_MAYBE in favor of KJ_IF_SOME. Please use KJ_IF_SOME.
+
+#if defined (KJ_DEPRECATE_EMPTY_MAYBE_FROM_NULLPTR) && KJ_DEPRECATE_EMPTY_MAYBE_FROM_NULLPTR
+#define KJ_DEPRECATE_EMPTY_MAYBE_FROM_NULLPTR_ATTR \
+    [[deprecated("Using nullptr as an empty Maybe is deprecated and will be removed. " \
+      "Please use kj::none for this purpose instead.")]]
+#else
+#define KJ_DEPRECATE_EMPTY_MAYBE_FROM_NULLPTR_ATTR
+#endif
+// Originally, we used `nullptr` to mean "an empty Maybe" in comparisons and initializations. For
+// example, `maybe == nullptr` would be true if `maybe` contained no value, and `Maybe<T>(nullptr)`
+// could be used to explicitly initialize an empty Maybe. This isn't the best design, because
+// `nullptr` could itself be a valid constructor parameter for the wrapped type T.
+//
+// Due to this flaw, we've deprecated using `nullptr` to mean "an empty Maybe" in favor of
+// `kj::none`. Please use `kj::none`.
+
+#if __GNUC__
+// Both clang and GCC define __GNUC__ and understand the GCC set of pragma directives.
```

Comment:
I believe that clang does NOT define __GNUC__ when operating in MSVC-compatible mode, but I suspect it'll still accept the GCC-style pragmas in that case.

Reply:
Thanks, switched to `__GNUC__ || __clang__`: https://github.com/capnproto/capnproto/compare/b8f0d95ca9fffa2dd374e746b7759957cb8ac1bd..2818ef1435057c62401ec6fbb7b50c65afee7957